src/assets/src/components/common.tsx

```diff
@@ -290,25 +245,93 @@ export const BackendSelector: React.FC<BackendSelectorProps> = (props) => {
     );
 }
 
-interface SingleInputFormShowRemainingProps extends StatelessSingleInputFormProps {
-    maxLength: number
+
+interface ValidatedInputFormProps extends SingleInputFormProps {
+    maxLength?: number;
+    validateLength?: boolean;
+    allowBlank: boolean;
 }
 
-export const SingleInputFormShowRemaining: React.FC<SingleInputFormShowRemainingProps> = (props) => {
-    const buttonClass = `btn btn-${props.buttonType} remaining-controls`;
-    const remaining = props.maxLength - props.value.length;
-    const isInvalid = props.value.length > props.maxLength;
+interface StatelessValidatedInputFormProps extends ValidatedInputFormProps {
+    value: string;
+    onChangeValue: (value: string) => void;
+}
+
+function generateFeedbackText (maxLength: number, currentLength: number, allowBlank: boolean) {
+    const remaining = maxLength - currentLength;
+    const charsRemaining = (remaining > 0) ? remaining : 0;
+    if (!allowBlank && currentLength === 0) {
+       return 'This field is not allowed to be blank.';
+    } else {
+        const charsOver = (remaining < 0) ? ` (${remaining * -1} over limit)` : '';
+        return `Remaining characters: ${charsRemaining}/${maxLength}${charsOver}`;
+    }
+}
+
+export const StatelessInputGroupForm: React.FC<StatelessValidatedInputFormProps> = (props) => {
```

Comment:
It looks like the point here is to provide a wrapper for a single arbitrary input element and any associated feedback? It might get complicated if we want to have a select or a type-ahead instead of a regular input (thinking ahead to replacing uniqname inputs with typeahead widgets).

Reply:
I think the idea is that the single input form is a common UI feature of the application. I agree that those future field types wouldn't be easily handled by this, but I think there are ways we continue to modularize over time. I think this is at least a progressive improvement?